framework/decode/vulkan_address_replacer.cpp

```diff
@@ -0,0 +1,791 @@
+/*
+** Copyright (c) 2024 LunarG, Inc.
+**
+** Permission is hereby granted, free of charge, to any person obtaining a
+** copy of this software and associated documentation files (the "Software"),
+** to deal in the Software without restriction, including without limitation
+** the rights to use, copy, modify, merge, publish, distribute, sublicense,
+** and/or sell copies of the Software, and to permit persons to whom the
+** Software is furnished to do so, subject to the following conditions:
+**
+** The above copyright notice and this permission notice shall be included in
+** all copies or substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+** FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+** AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+** FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+** DEALINGS IN THE SOFTWARE.
+*/
+
+#include "decode/vulkan_address_replacer.h"
+#include "decode/vulkan_address_replacer_shaders.h"
+#include "decode/mark_injected_commands.h"
+#include "util/logging.h"
+
+GFXRECON_BEGIN_NAMESPACE(gfxrecon)
+GFXRECON_BEGIN_NAMESPACE(decode)
+
+//! RAII helper to mark injected commands in scope
+struct mark_injected_commands_helper_t
+{
+    mark_injected_commands_helper_t()
+    {
+        // mark injected commands
+        decode::BeginInjectedCommands();
+    }
+
+    ~mark_injected_commands_helper_t()
+    {
+        // mark end of injected commands
+        decode::EndInjectedCommands();
+    }
+};
+
+inline uint32_t aligned_size(uint32_t size, uint32_t alignment)
+{
+    return (size + alignment - 1) & ~(alignment - 1);
+}
+
+inline uint32_t div_up(uint32_t nom, uint32_t denom)
+{
+    GFXRECON_ASSERT(denom > 0)
+    return (nom + denom - 1) / denom;
+}
+
+uint32_t get_memory_type_index(const VkPhysicalDeviceMemoryProperties& memory_properties,
+                               uint32_t                                type_bits,
+                               VkMemoryPropertyFlags                   property_flags)
+{
+    uint32_t memory_type_index = std::numeric_limits<uint32_t>::max();
+
+    for (uint32_t i = 0; i < memory_properties.memoryTypeCount; ++i)
+    {
+        if ((type_bits & (1 << i)) &&
+            ((memory_properties.memoryTypes[i].propertyFlags & property_flags) == property_flags))
+        {
+            memory_type_index = i;
+            break;
+        }
+    }
+
+    return memory_type_index;
+}
+
+//! 16 bytes
```

Comment:
What's the point of this comment?

Reply:
was me, counting bytes during development. I'll remove it :)

Reply:
done